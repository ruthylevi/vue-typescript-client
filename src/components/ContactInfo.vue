<template>
  <v-hover v-model="hovered">
    <v-card height="511">
      <v-container>
        <v-container style="height: 450px;">
          <v-data-iterator
            :items="members"
            :items-per-page="itemsPerPage"
            :page.sync="page"
            hide-default-footer
          >
            <template v-slot:default="props">
              <v-slide-x-transition group hide-on-leave>
                <v-row v-for="item in props.items" :key="item.id" align="center" class="spacer">
                  <v-col cols="4" sm="2" md="1">
                    <v-tooltip v-if="isAdmin(item.id) && isAdvisor(item.id)" top max-width="175">
                      <template v-slot:activator="{ on }">
                        <v-badge
                          color="yellow darken-2"
                          icon="mdi-crown"
                          overlap
                          offset-x="18"
                          offset-y="7"
                        >
                          <v-badge
                            color="purple darken-2"
                            icon="mdi-school"
                            overlap
                            offset-x="7"
                            offset-y="18"
                          >
                            <v-avatar
                              size="36"
                              :color="item.thumbnailLink !== null ? undefined : 'primary'"
                              v-on="on"
                            >
                              <v-img v-if="item.thumbnailLink !== null" :src="item.thumbnailLink" />
                              <v-icon v-else dark>mdi-account-circle</v-icon>
                            </v-avatar>
                          </v-badge>
                        </v-badge>
                      </template>
                      <span>Administrator and advisor of this project.</span>
                    </v-tooltip>
                    <v-tooltip v-else-if="isAdmin(item.id)" top max-width="175">
                      <template v-slot:activator="{ on }">
                        <v-badge color="yellow darken-2" icon="mdi-crown" overlap>
                          <v-avatar
                            size="36"
                            :color="item.thumbnailLink !== null ? undefined : 'primary'"
                            v-on="on"
                          >
                            <v-img v-if="item.thumbnailLink !== null" :src="item.thumbnailLink" />
                            <v-icon v-else dark>mdi-account-circle</v-icon>
                          </v-avatar>
                        </v-badge>
                      </template>
                      <span>Administrator of this project.</span>
                    </v-tooltip>
                    <v-tooltip v-else-if="isAdvisor(item.id)" top max-width="175">
                      <template v-slot:activator="{ on }">
                        <v-badge color="purple darken-2" icon="mdi-school" overlap>
                          <v-avatar
                            size="36"
                            :color="item.thumbnailLink !== null ? undefined : 'primary'"
                            v-on="on"
                          >
                            <v-img v-if="item.thumbnailLink !== null" :src="item.thumbnailLink" />
                            <v-icon v-else dark>mdi-account-circle</v-icon>
                          </v-avatar>
                        </v-badge>
                      </template>
                      <span>Advisor of this project.</span>
                    </v-tooltip>
                    <v-avatar
                      v-else
                      size="36"
                      :color="item.thumbnailLink !== null ? undefined : 'primary'"
                    >
                      <v-img v-if="item.thumbnailLink !== null" :src="item.thumbnailLink" />
                      <v-icon v-else dark>mdi-account-circle</v-icon>
                    </v-avatar>
                  </v-col>
                  <v-col>
                    <v-list-item-title class="ml-7 text-left" style="max-width: 266px;">
                      {{ item.firstName }} {{ item.lastName }}
                    </v-list-item-title>
                    <v-list-item-subtitle
                      class="ml-7 text-left grey--text"
                      style="max-width: 266px;"
                    >
                      {{ item.email }}
                    </v-list-item-subtitle>
                  </v-col>
                </v-row>
              </v-slide-x-transition>
            </template>
          </v-data-iterator>
        </v-container>
        <v-row v-if="numberOfPages !== 1" class="mx-2" align="center" justify="center">
          <v-btn
            icon
            dark
            color="secondary"
            class="mx-1"
            @click="formerPage()"
            :disabled="page === 1"
          >
            <v-icon>mdi-chevron-left</v-icon>
          </v-btn>
          <span class="mx-1 secondary--text"> Page {{ page }}/{{ numberOfPages }} </span>
          <v-btn
            icon
            dark
            color="secondary"
            class="mx-1"
            @click="nextPage()"
            :disabled="page === numberOfPages"
          >
            <v-icon>mdi-chevron-right</v-icon>
          </v-btn>
        </v-row>
      </v-container>
    </v-card>
  </v-hover>
</template>

<script>
export default {
  props: {
    contributors: Array,
    administrators: Array,
  },
  data() {
    return {
      page: 1,
      itemsPerPage: 6,
      hovered: true,
    };
  },
  methods: {
    isContributor(personId) {
      if (this.members === undefined) return [];
      return this.contributors.map(admin => admin.id).includes(personId);
    },
    isAdmin(personId) {
      if (this.members === undefined) return [];
      return this.administrators.map(admin => admin.id).includes(personId);
    },
    isAdvisor(personId) {
      if (this.members === undefined) return [];
      return this.members
        .filter(member => member.roles.includes('faculty'))
        .map(admin => admin.id)
        .includes(personId);
    },
    nextPage() {
      if (this.page + 1 <= this.numberOfPages) this.page += 1;
    },
    formerPage() {
      if (this.page - 1 >= 1) this.page -= 1;
    },
  },
  mounted() {
    const interval = setInterval(() => {
      if (!this.hovered && this.numberOfPages > 1) {
        if (this.page + 1 <= this.numberOfPages) {
          this.page += 1;
        } else {
          this.page = 1;
        }
      }
    }, 3000);
  },
  computed: {
    members() {
      if (this.administrators === undefined || this.contributors === undefined) {
        return [];
      }
      return this.administrators.concat(this.contributors);
    },
    numberOfPages() {
      return Math.ceil(this.members.length / this.itemsPerPage);
    },
  },
};
</script>
