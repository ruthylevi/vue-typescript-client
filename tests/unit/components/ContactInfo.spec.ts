import Vue from 'vue';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';
import { shallowMount } from '@vue/test-utils';
import ContactInfo from '@/components/ContactInfo.vue';

Vue.use(Vuetify);
Vue.use(VueRouter);

const wrapper = shallowMount(ContactInfo);

describe('ContactInfo', () => {
  test('Sanity', () => {
    expect(true).toBe(true);
  });

  test('Visible', () => {
    expect(wrapper.isVisible()).toBe(true);
  });

  test('Renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
