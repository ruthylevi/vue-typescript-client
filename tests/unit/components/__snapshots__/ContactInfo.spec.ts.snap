// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInfo Renders correctly 1`] = `
<v-card-stub loaderheight="4" tag="div">
  <v-container-stub tag="div">
    <h2>Advisors:</h2>
    <v-container-stub tag="div"></v-container-stub>
    <h2>Members:</h2>
    <v-container-stub tag="div"></v-container-stub>
  </v-container-stub>
</v-card-stub>
`;
